### PR TITLE
refactor(general): drop pydantic v1

### LINKED ISSUE #4191: Maintenance: Remove any specific code for Pydantic v1
### Why is this needed?

We are working to release Powertools for AWS Lambda (Python) v3 and in this new version we will remove support for Pydantic v1, keeping Pydantic v2 as the only supported version. Our code and tests are full of `ifs` to ensure both versions work and we need to remove this.

We also need to bump the Pydantic version in `pyproject.toml`

### Which area does this relate to?

Other

### Solution

_No response_

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by leandrodamascena**
**Issue number:** #4191 

### Changes

This pull request removes support for Pydantic v1 in Powertools for AWS Lambda (Python) v3. Pydantic v2 will be the only supported version going forward.

Our codebase and test suite currently have many conditional checks to ensure compatibility with both Pydantic v1 and v2. This makes the codebase more complex and harder to maintain. By removing support for Pydantic v1, we can simplify the codebase and focus on providing the best experience for Pydantic v2 users.

Customers using Powertools for AWS Lambda (Python) v2 will not need to update their codebase, as those versions will continue to support Pydantic v1. However, customers upgrading to v3 will need to ensure that their codebase uses Pydantic v2. This may require updating their dependencies and making any necessary changes to their code to be compatible with the Pydantic v2 API.

### User experience

> Please share what the user experience looks like before and after this change

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

File: aws_lambda_powertools/utilities/parser/models/s3_batch_operation.py [hunk 1]
```diff
@@ -12,7 +12,7 @@ class S3BatchOperationTaskModel(BaseModel):
     s3BucketArn: Optional[str] = None
     s3Bucket: Optional[str] = None
 
-    @validator("s3Bucket", pre=True, always=True)
+    @validator("s3Bucket", pre=True, always=True)  # review
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/s3_batch_operation.py]: NOTE: need to review before merge.

File: aws_lambda_powertools/utilities/parser/models/s3.py [hunk 2]
```diff
@@ -101,7 +101,7 @@ class S3RecordModel(BaseModel):
     s3: S3Message
     glacierEventData: Optional[S3EventRecordGlacierEventData] = None
 
-    @root_validator(allow_reuse=True, skip_on_failure=True)
+    @root_validator(allow_reuse=True, skip_on_failure=True)  # review
```
Comment (leandrodamascena) [inline aws_lambda_powertools/utilities/parser/models/s3.py]: NOTE: need to review before merge.

File: tests/functional/event_handler/test_openapi_encoders.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 import pytest
 from pydantic import BaseModel
-from pydantic.color import Color
+from pydantic.color import Color  # review
```
Comment (leandrodamascena) [inline tests/functional/event_handler/test_openapi_encoders.py]: NOTE: need to review before merge.

File: aws_lambda_powertools/utilities/parser/models/s3_batch_operation.py (hunk 1 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/models/s3_batch_operation.py]: Fixed

File: aws_lambda_powertools/utilities/parser/models/s3.py (hunk 2 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/utilities/parser/models/s3.py]: Fixed

**Comment by leandrodamascena** [comment]
Hey @heitorlessa, this PR is ready to review! We may need to discuss about this: https://github.com/aws-powertools/powertools-lambda-python/pull/4305/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R99.

I can't see any Pydantic deprecation warning in the tests, but please double check this.

File: aws_lambda_powertools/event_handler/openapi/compat.py [hunk 4]
```diff
@@ -15,13 +15,27 @@
...
 
-from aws_lambda_powertools.event_handler.openapi.pydantic_loader import PYDANTIC_V2
 from aws_lambda_powertools.event_handler.openapi.types import (
     COMPONENT_REF_PREFIX,
     ModelNameMap,
     UnionType,
 )
 
+from pydantic import TypeAdapter, ValidationError
+from pydantic._internal._typing_extra import eval_type_lenient
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/compat.py:29]: is there another more stable way? 

`_internal` is private so this **might** cause breaking changes. If there's no other way, we need to make a comment of the risk

File: aws_lambda_powertools/event_handler/openapi/models.py [hunk 5]
```diff
@@ -19,12 +18,7 @@ class Contact(BaseModel):
     url: Optional[AnyUrl] = None
     email: Optional[str] = None
 
-    if PYDANTIC_V2:
-        model_config = {"extra": "allow"}
-    else:
-
-        class Config:
-            extra = "allow"
+    model_config = {"extra": "allow"}
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/models.py]: can we perhaps define it as a top-level constant to reduce memory usage?

It looks like we always allow extra keys for these models

File: aws_lambda_powertools/event_handler/openapi/swagger_ui/oauth2.py [hunk 6]
```diff
@@ -42,15 +41,9 @@ class OAuth2Config(BaseModel):
     # Whether to use PKCE with the authorization code grant type. Defaults to False.
     usePkceWithAuthorizationCodeGrant: bool = Field(alias="use_pkce_with_authorization_code_grant", default=False)
 
-    if PYDANTIC_V2:
-        model_config = {"extra": "allow"}
-    else:
+    model_config = {"extra": "allow"}
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/swagger_ui/oauth2.py]: same as earlier about a constant. a `openapi/constants.py` that has a model config to allow extra keys

File: aws_lambda_powertools/event_handler/openapi/types.py [hunk 7]
```diff
@@ -41,7 +41,7 @@
         "detail": {
             "title": "Detail",
             "type": "array",
-            "items": {"$ref": COMPONENT_REF_PREFIX + "ValidationError"},
+            "items": {"$ref": f"{COMPONENT_REF_PREFIX}ValidationError"},
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/types.py:44]: great catch to optimize memory! <3

File: aws_lambda_powertools/utilities/parser/parser.py [hunk 8]
```diff
@@ -2,13 +2,11 @@
 import typing
 from typing import Any, Callable, Dict, Optional, Type, overload
 
-from aws_lambda_powertools.utilities.parser.compat import disable_pydantic_v2_warning
+from aws_lambda_powertools.middleware_factory import lambda_handler_decorator
```
Comment (heitorlessa) [inline aws_lambda_powertools/utilities/parser/parser.py:5]: NICE change from relative to absolute import <3

File: aws_lambda_powertools/event_handler/openapi/constants.py [hunk 9]
```diff
@@ -1,2 +1,2 @@
 DEFAULT_API_VERSION = "1.0.0"
-DEFAULT_OPENAPI_VERSION = "3.0.3"
+DEFAULT_OPENAPI_VERSION = "3.1.0"
```
Comment (heitorlessa) [inline aws_lambda_powertools/event_handler/openapi/constants.py:4]: do we already have the Bedrock Agent workaround as their docs are wrong?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
superb work (I know how exhaustive this is) -- left one question about Bedrock Agents OpenAPI version workaround, and one ask to reduce memory usage for `model_config = {"extra": "allow"}` (define once).

**Comment by heitorlessa** [comment]
Random question, do you know the reason why we have this? https://github.com/aws-powertools/powertools-lambda-python/blob/17b62b8ffeee546c6051a55661c3e9932a92e79c/aws_lambda_powertools/event_handler/openapi/encoders.py#L321

It's the only place where we'd need this extra dependency.

File: aws_lambda_powertools/event_handler/openapi/swagger_ui/oauth2.py (hunk 6 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/swagger_ui/oauth2.py]: Yeah, make sense! Done.

File: aws_lambda_powertools/event_handler/openapi/models.py (hunk 5 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/models.py]: Yeah, make sense! Done.

File: aws_lambda_powertools/event_handler/openapi/constants.py (hunk 9 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/constants.py:4]: We will resolve issues with Agent Bedrock in this other PR @heitorlessa

https://github.com/aws-powertools/powertools-lambda-python/issues/4196

File: aws_lambda_powertools/event_handler/openapi/compat.py (hunk 4 above)
Comment (leandrodamascena) [reply aws_lambda_powertools/event_handler/openapi/compat.py:29]: Good catch! Added a comment

**Review by heitorlessa — APPROVED** [review]
LGTM!!!!

**Pull request merged**